File: BaekJoon/Silver/15663/Python/15663.py
# no. 15663: N과 M (9)
# 백트래킹
# 같은 숫자를 여러번 뽑아도 되지만, 중복되는 수열은 출력X
# 4 4 2 같은 수열이 가능하다는 뜻

import sys
input = sys.stdin.readline

n, m = map(int, input().split())
nums = {}
for num in map(int, input().split()):
    if num not in nums:
        nums[num] = 1
    else:
        nums[num] += 1

def dfs(depth=m, cur_sequence=[], sequences=[]):
    if depth == 0:
        sequences.append(cur_sequence.copy())
        cur_sequence.pop()
        return
    for num in nums:
        if cur_sequence.count(num) == nums[num]:
            continue
        cur_sequence.append(num)
        dfs(depth-1, cur_sequence, sequences)
    if cur_sequence:
        cur_sequence.pop()
    return sequences

seq = dfs()
seq.sort()
for arr in seq:
    print(*arr)

File: BaekJoon/Silver/15663/C++/15663.cpp
// no.15663: N과 M (9) (S2)

#include <cstdio>
#include <algorithm>
#include <vector>
using namespace std;

void dfs(int, int, int, vector<int>, vector<vector<int>>&);

int n, m;
vector<int> nums;

int main() {
    scanf("%d %d", &n, &m);
    for(int i=0; i<n; i++) {
        int num;
        scanf("%d", &num);
        nums.push_back(num);
    }
    sort(nums.begin(), nums.end());
    vector<vector<int>> ret;
    dfs(-1, 0, 0, {}, ret);
    for(vector<int>& arr : ret) {
        for(int num : arr) {
            printf("%d ", num);
        }
        printf("\n");
    }
    return 0;
}

void dfs(int idx, int depth, int check, vector<int> tmp, vector<vector<int>>& ret) {
    if(idx != -1) {
        tmp.push_back(nums[idx]);
        check = check | (1<<idx);
    }
    if(depth == m) {
        ret.push_back(tmp);
        return;
    }
    vector<int> selected;
    for(int i=0; i<n; i++) {
        if((1<<i) & check)
            continue;
        if(find(selected.begin(), selected.end(), nums[i]) != selected.end())
            continue;
        selected.push_back(nums[i]);
        dfs(i, depth+1, check, tmp, ret);
    }
}